pymc_experimental/marginal_model.py

```diff
@@ -478,3 +498,64 @@ def logp_fn(marginalized_rv_const, *non_sequences):
 
     # We have to add dummy logps for the remaining value variables, otherwise PyMC will raise
     return joint_logps, *(pt.constant(0),) * (len(values) - 1)
+
+
+@_logprob.register(DiscreteMarginalMarkovChainRV)
+def finite_discrete_marginal_rv_logp(op, values, *inputs, **kwargs):
+    def step_alpha(log_alpha, log_P):
+        return pt.logsumexp(log_alpha[:, None] + log_P, 0)
+
+    def eval_logp(x):
+        return logp(init_dist_, x)
+
+    marginalized_rvs_node = op.make_node(*inputs)
+    inner_rvs = clone_replace(
+        op.inner_outputs,
+        replace={u: v for u, v in zip(op.inner_inputs, marginalized_rvs_node.inputs)},
+    )
+
+    chain_rv, *dependent_rvs = inner_rvs
+    P_, n_steps_, init_dist_, rng = chain_rv.owner.inputs
+
+    domain = pt.arange(P_.shape[0], dtype="int32")
+
+    vec_eval_logp = pt.vectorize(eval_logp, "()->()")
+    logp_init = vec_eval_logp(domain)
+
+    # Construct logp in two steps
+    # Step 1: Compute the probability of the data ("emissions") under every possible state (vec_logp_emission)
+
+    # This will break the dependency between chain and the init_dist_ random variable
+    # TODO: Make this comment more robust after I understand better.
+    chain_dummy = chain_rv.clone()
+    dependent_rvs = clone_replace(dependent_rvs, {chain_rv: chain_dummy})
+    input_dict = dict(zip(dependent_rvs, values))
+    logp_value_dict = conditional_logp(input_dict)
+
+    # TODO: Is values[0] robust to every situation?
+    sub_dict = {
+        chain_dummy: pt.moveaxis(pt.broadcast_to(domain, (*values[0].shape, domain.size)), -1, 0)
+    }
+
+    # This is a (k, T) matrix of logp terms, one for each state - observation pair
+    vec_logp_emission = vectorize_graph(tuple(logp_value_dict.values()), sub_dict)
+
+    # Step 2: Compute the transition probabilities
+    # This is the "forward algorithm", alpha_t = sum(p(s_t | s_{t-1}) * alpha_{t-1})
+    # We do it entirely in logs, though.
+    log_alpha_seq, _ = scan(
+        step_alpha, non_sequences=[pt.log(P_)], outputs_info=[logp_init], n_steps=n_steps_
+    )
+
+    # Scan works over the T dimension, so output is (T, k). We need to swap to (k, T)
+    log_alpha_seq = pt.moveaxis(pt.concatenate([logp_init[None], log_alpha_seq], axis=0), -1, 0)
+
+    # Final logp is the sum of the sum of the emission probs and the transition probabilities
+    # pt.add is used in case there are multiple emissions that depend on the same markov chain; in this case, we compute
+    # the joint probability of seeing everything together.
+    joint_log_obs_given_states = pt.logsumexp(pt.add(*vec_logp_emission) + log_alpha_seq, axis=0)
+
+    # If there are multple emisson streams, we have to add dummy logps for the remaining value variables. The first
+    # return is the joint probability of everything together, but PyMC still expects one logp for each one.
+    dummy_logps = (pt.constant(np.zeros((4,))),) * (len(values) - 1)
```

Comment:
Does it work with empty arrays? That would be "mathematically cleaner". Same for the pre-existing marginal discrete logp
```suggestion
    dummy_logps = (pt.constant(np.zeros((0,))),) * (len(values) - 1)
```

Reply:
This is debug code that I left in my accident. I'll make a test case for empty arrays